### PR TITLE
added profit and loss api controller and test class

### PR DESCRIPTION
**Pull request by faisal-amin82**
- added profit and loss api controller and created test class also updated SmallFullLink class

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/ProfitLossController.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import uk.gov.companieshouse.api.accounts.utility.ApiResponseMapper;
+import uk.gov.companieshouse.api.accounts.utility.ErrorMapper;
+import uk.gov.companieshouse.api.accounts.utility.LoggingHelper;
+import uk.gov.companieshouse.api.model.transaction.Transaction;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.validation.Valid;
+
+@RestController
+@RequestMapping(value = "/transactions/{transactionId}/company-accounts/{companyAccountId}/small-full/notes/profit-and-loss", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/ProfitLossController.java]: This isn't a note - it's a child of current or previous period. We could achieve this functionality using a single controller by using a `RequestMapping` value of:
```
/transactions/{transactionId}/company-accounts/{companyAccountId}/small-full/{period:(current-period|previous-period)}/profit-and-loss
```

The `period` variable would be a string which could be used for logic, particularly in the `PUT` mapped method as explained below. We could even capture `period` in an enum for readability; see [here](https://www.devglan.com/spring-boot/enums-as-request-parameters-in-spring-boot-rest) for one example on how to achieve this

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/ProfitLossController.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    private ProfitLossService profitLossService;
+
+    @Autowired
+    private ApiResponseMapper apiResponseMapper;
+
+    @Autowired
+    private ErrorMapper errorMapper;
+
+    @PostMapping
+    public ResponseEntity create(@Valid @RequestBody ProfitLoss profitLoss,
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/ProfitLossController.java]: Something I didn't consider, the `ProfitLoss` object will need to be refactored to become `ProfitAndLoss`. When Jackson serializes an object, the object name is snake-cased and prefixed with a dollar to create the prefix of the json path.

As the object is currently named, Spring's binding result errors will result in json of the form:
```
$.profit_loss.<path_of_field_in_error>
```

Renaming this object to Profit**And**Loss will correctly yield validation errors of the format:
```
$.profit_and_loss.<path_of_field_in_error>
```

File: src/main/java/uk/gov/companieshouse/api/accounts/links/SmallFullLinkType.java [hunk 3]
```diff
@@ -19,7 +19,8 @@
     FIXED_ASSETS_INVESTMENTS_NOTE("fixed_assets_investments_note"),
     EMPLOYEES_NOTE("employees_note"),
     CURRENT_PERIOD_PROFIT_LOSS("current_period_profit_and_loss"),
-    PREVIOUS_PERIOD_PROFIT_LOSS("previous_period_profit_and_loss");
+    PREVIOUS_PERIOD_PROFIT_LOSS("previous_period_profit_and_loss"),
+    PROFIT_AND_LOSS("profit_and_loss");
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/links/SmallFullLinkType.java]: This doesn't need to be here, reference either the current or previous period variant of profit and loss

File: src/main/java/uk/gov/companieshouse/api/accounts/links/CurrentPeriodLinkType.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.companieshouse.api.accounts.links;
+
+public enum CurrentPeriodLinkType implements LinkType {
+
+    SELF("self"),
+    APPROVAL("approval"),
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/links/CurrentPeriodLinkType.java]: I don't think this is needed?

File: src/main/java/uk/gov/companieshouse/api/accounts/links/PreviousPeriodLinkType.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.companieshouse.api.accounts.links;
+
+public enum PreviousPeriodLinkType implements LinkType {
+
+    SELF("self"),
+    APPROVAL("approval"),
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/links/PreviousPeriodLinkType.java]: Nor here

File: src/main/java/uk/gov/companieshouse/api/accounts/links/SmallFullLinkType.java (hunk 3 above)
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/links/SmallFullLinkType.java]: No profit and loss references are required in this class

File: src/main/java/uk/gov/companieshouse/api/accounts/service/impl/CurrentPeriodService.java [hunk 6]
```diff
@@ -130,6 +134,40 @@ public CurrentPeriodService(
...
+
+    }
+
+    @Override
+    public void removeLink(String id, CurrentPeriodLinkType linkType, HttpServletRequest request) throws DataException {
+
+        String currentPeriodId = generateID(id);
+        CurrentPeriodEntity currentPeriodEntity = currentPeriodRepository.findById(currentPeriodId)
+                .orElseThrow(() -> new DataException(
+                        "Failed to get profit and loss entity from which to remove link"));
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/service/impl/CurrentPeriodService.java]: Same here

File: src/main/java/uk/gov/companieshouse/api/accounts/service/impl/PreviousPeriodService.java [hunk 7]
```diff
@@ -111,6 +115,40 @@ public PreviousPeriodService(
...
     }
 
+    @Override
+    public void addLink(String id, PreviousPeriodLinkType linkType, String link, HttpServletRequest request)
+            throws DataException {
+        String previousPeriodId = generateID(id);
+        try {
+        PreviousPeriodEntity previousPeriodEntity = previousPeriodRepository.findById(previousPeriodId)
+                .orElseThrow(() -> new DataException(
+                        "Failed to get profit and loss entity to add link"));
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/service/impl/PreviousPeriodService.java]: As above

File: src/main/java/uk/gov/companieshouse/api/accounts/service/impl/PreviousPeriodService.java [hunk 8]
```diff
@@ -111,6 +115,40 @@ public PreviousPeriodService(
...
+    }
+
+    @Override
+    public void removeLink(String id, PreviousPeriodLinkType linkType, HttpServletRequest request)
+            throws DataException {
+
+        String smallFullId = generateID(id);
+        PreviousPeriodEntity previousPeriodEntity = previousPeriodRepository.findById(smallFullId)
+                .orElseThrow(() -> new DataException(
+                        "Failed to get profit and loss entity from which to remove link"));
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/service/impl/PreviousPeriodService.java]: And here

File: src/test/java/uk/gov/companieshouse/api/accounts/interceptor/CurrentPeriodInterceptorTest.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    private CurrentPeriodInterceptor currentPeriodInterceptor;
+
+    @BeforeEach
+    public void setUp() throws NoSuchAlgorithmException {
+
+        Map<String, String> pathVariables = new HashMap<>();
+        pathVariables.put("transactionId", "5555");
+        pathVariables.put("companyAccountId", "test");
+
+        when(httpServletRequest.getHeader("X-Request-Id")).thenReturn("test");
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/api/accounts/interceptor/CurrentPeriodInterceptorTest.java]: Can we remove magic strings please? Define `private static final String`'s for constants

File: src/test/java/uk/gov/companieshouse/api/accounts/service/impl/CurrentPeriodServiceTest.java [hunk 10]
```diff
@@ -62,12 +74,24 @@
...
     private Errors errors;
 
+    @Mock
+    private SmallFull smallFull;
+
+    @Mock
+    private Map<String, String> links;
+
+    @Mock
+    CurrentPeriodController currentPeriodController;
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/api/accounts/service/impl/CurrentPeriodServiceTest.java]: Is this required?

File: src/main/java/uk/gov/companieshouse/api/accounts/links/PreviousPeriodLinkType.java (hunk 5 above)
Comment (faisal-amin82) [reply src/main/java/uk/gov/companieshouse/api/accounts/links/PreviousPeriodLinkType.java]: removed as suggested

File: src/main/java/uk/gov/companieshouse/api/accounts/links/SmallFullLinkType.java (hunk 3 above)
Comment (faisal-amin82) [reply src/main/java/uk/gov/companieshouse/api/accounts/links/SmallFullLinkType.java]: removed as suggested

File: src/test/java/uk/gov/companieshouse/api/accounts/service/impl/CurrentPeriodServiceTest.java (hunk 10 above)
Comment (faisal-amin82) [reply src/test/java/uk/gov/companieshouse/api/accounts/service/impl/CurrentPeriodServiceTest.java]: Removed as it was not needed

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/Period.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package uk.gov.companieshouse.api.accounts.controller;
+
+import java.util.Arrays;
+
+public enum Period {
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/Period.java]: Can we rename this to `AccountingPeriod` to be more explicit? Keep the private `period` field named as is - Sonar doesn't like conflicting field and object names

File: src/test/java/uk/gov/companieshouse/api/accounts/interceptor/CurrentPeriodInterceptorTest.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class CurrentPeriodInterceptorTest {
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/api/accounts/interceptor/CurrentPeriodInterceptorTest.java]: Code coverage on this class could be improved

File: src/test/java/uk/gov/companieshouse/api/accounts/interceptor/PreviousPeriodInterceptorTest.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+
+@ExtendWith(MockitoExtension.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+  class PreviousPeriodInterceptorTest {
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/api/accounts/interceptor/PreviousPeriodInterceptorTest.java]: As above

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/Period.java (hunk 11 above)
Comment (faisal-amin82) [reply src/main/java/uk/gov/companieshouse/api/accounts/controller/Period.java]: renamed Period  class to AccountingPeriod

**Pull request merged**